lib/menubar/menu_gen.lua

```diff
@@ -83,48 +83,60 @@ local function trim(s)
-                            local cat_key, cat_use =
-                            get_category_name_and_usage_by_type(category)
-                            if cat_key and cat_use then
-                                target_category = cat_key
-                                break
+        utils.parse_dir(dir, function(entries)
+            if not entries then
+                dirs_parsed = dirs_parsed + 1
+                return
+            end
+            for _, entry in ipairs(entries) do
+                -- Check whether to include program in the menu
+                if entry.show and entry.Name and entry.cmdline then
+                    local unique_key = entry.Name .. '\0' .. entry.cmdline
+                    if not unique_entries[unique_key] then
+                        local target_category = nil
+                        -- Check if the program falls into at least one of the
+                        -- usable categories. Set target_category to be the id
+                        -- of the first category it finds.
+                        if entry.categories then
+                            for _, category in pairs(entry.categories) do
+                                local cat_key, cat_use =
+                                get_category_name_and_usage_by_type(category)
+                                if cat_key and cat_use then
+                                    target_category = cat_key
+                                    break
+                                end
                             end
                         end
-                    end
-                    if target_category then
-                        local name = trim(entry.Name) or ""
-                        local cmdline = trim(entry.cmdline) or ""
-                        local icon = entry.icon_path or nil
-                        table.insert(result, { name = name,
-                                     cmdline = cmdline,
-                                     icon = icon,
-                                     category = target_category })
-                        unique_entries[unique_key] = true
+                        if target_category then
+                            local name = trim(entry.Name) or ""
+                            local cmdline = trim(entry.cmdline) or ""
+                            local icon = entry.icon_path or nil
+                            table.insert(result, { name = name,
+                                         cmdline = cmdline,
+                                         icon = icon,
+                                         category = target_category })
+                            unique_entries[unique_key] = true
+                        end
                     end
                 end
             end
-        end
+            dirs_parsed = dirs_parsed + 1
+            if dirs_parsed == #menu_gen.all_menu_dirs then
+                callback(result)
+            end
```

Comment:
This part is missing on the initial `if not entries then dirs_parsed = dirs_parsed + 1`.
I suggest replacing this initial check with `entries = entries or {}`. That would get rid of this special case.